Fix registering jit decompositions for jvp for out wrapped decomps

kundaMwiza opened the pull request:
Python decompositions wrapped by `out_wrapper` need to be unwrapped before compiling with TorchScript since:
- `out_wrapper` extends the decompositions signature with an out parameter, however this `out` parameter is not present in the source code of the original decomposition so the resulting `ScriptFunction` will not have an `out` parameter
- `out_wrapper` is in the `torch._prims_common.wrappers` module so its `globals()` are different to the globals of the decomposition to be wrapped. This may cause symbol resolution to fail with the TorchScript compiler since it is compiling the unwrapped decomps source code rather than the wrapper

The python decomposition for `aten.trace` is wrapped as an example, other decompositions are to be fixed in https://github.com/pytorch/pytorch/pull/107707

kundaMwiza committed:
Add attribute to out_wrapper to indicate that the function is wrapped by it. Jit tests are currently incompatible with out_wrapper so this attribute can be used to unwrap decomps wrapped by out_wrapper

Ignore mypy attr-defined error and lint

Dynamically resolve location of out_wrapper

Create a convenient function for removing the out wrapper

Clarify comment for removing the out wrapper to trace. prefix maybe_remove_out_wrapper with an underscore

Lint

Move unwrapping of decomps to test_out_warning instead as that is where the error occurs

Revert changes to trace decomp

Remove unused import

Do not overwrite original OpInfo

unwrap ops in test_out that do not support out

kundaMwiza commented:
@pytorchbot label "topic: not user facing"

kundaMwiza commented:
Required for https://github.com/pytorch/pytorch/pull/107707

kundaMwiza commented:
@Chillee bump

lezcano reviewed (commented):
This doesn't feel right. Wouldn't it be possible to fix the test and teach TorchScript how to handle these decompositions?

In all fairness, I didn't even know that TorchScript was using decompositions...

kundaMwiza commented:
@lezcano 

These are the decompositions affected: 
https://github.com/pytorch/pytorch/blob/47e721422cfa16e958f119dd4d0fde409b6e5d95/torch/_decomp/decompositions_for_jvp.py#L293

For this fix, I followed the pattern used for TorchScript decorators: https://github.com/pytorch/pytorch/blob/47e721422cfa16e958f119dd4d0fde409b6e5d95/torch/jit/_script.py#L1363C1-L1368C74

I think having torchscript support decorators may be more involved. An option would be changing `out_wrapper` so that it keeps its source code rather than inheriting the python decomps source code. This would then mean that the TorchScript compiler would have to recurse into the decomp being wrapped and compile that however I don't know if the compiler code can do that

lezcano commented:
Could we do this as a wrapper around `out_wrapper`. Meaning, have a function that calls `out_wrapper` and it's just used for these JVP decompositions, so that we do not need to modify the code within `out_wrapper` to support TorchScript (as TorchScript is not currently being maintained anyway)

kundaMwiza commented:
@lezcano  

Sorry I don't understand what you mean. The problem here is that the `_register_jit_decomposition_for_jvp` uses pre-existing decomps, so some of these are already wrapped by `out_wrapper` (e.g. in decompositions.py), so what is needed is to unwrap them before they are compiled.
 
```python
def _register_jit_decomposition_for_jvp(decomp, use_python=False):
    if decomp in decomposition_table_for_jvp:
        decomposition_table_used = decomposition_table_for_jvp
    elif decomp in decomposition_table:
        decomposition_table_used = decomposition_table
    else:
        raise RuntimeError(f"could not find decomposition for {decomp}")
    decomp_fn = decomposition_table_used[decomp]

    # `out_wrapper` extends a decompositions signature with
    # an `out` parameter. However jit will use the unwrapped function's
    # signature instead so we need to unwrap here to prevent an error
    decomp_fn = _maybe_remove_out_wrapper(decomp_fn)
    ...
```

lezcano reviewed (commented):
Ah, I understand now! Fair enough.

Could we also apply  `_maybe_remove_out_wrapper` within the `test_out_warning` test, rather than applying it to all the decompositions?

lezcano commented:
The failures are real

lezcano reviewed (commented):
I am a bit confused. Shouldn't this fix some tests that are currently skipped/xfailed? If so, we should remove the relevant skips.

kundaMwiza commented:
@lezcano 
This is a fix for test failures in https://github.com/pytorch/pytorch/pull/107707 but I separated it out for clarity. 

On main, in torch/_refs/__init__.py, trace is not wrapped with `out_wrapper`. In this PR I have added the `out_wrapper` to trace. Without the fix of unwrapping decomps, `test_out` and `test_out_warning` would fail (as they are in #107707, since I've added wrapping to many decomps). Trace is used as an example of why this unwrapping is required but the general fix will allow a lot of the failures in  #107707 to be fixed

lezcano commented:
Oh, so the issue here is that these functions do not have an `out=` variant in Python but they do in ATen. Ugh, I see. I thought it was just an issue about some weird interaction with TorchScript. Anyway, let's merge this hacky fix. A proper fix would be to expose the `out=` kwarg in Python, but well...

lezcano commented:
@pytorchbot merge

lezcano commented:
next time, consider using ghstack to submit PRs, and use it to stack PRs like these two

kundaMwiza commented:
Is there a reason why some torch ops have `out=` and others don't? This seems like something that should be handled by codegen?

lezcano commented:
Yes, this may very well be a bug in the `autogen:` key within codegen.
